src/ingo_hash/poseidon_api.rs

```diff
@@ -130,21 +146,22 @@ impl DriverPrimitive<Hash, PoseidonInitializeParameters, &[u8], Vec<PoseidonResu
         todo!()
     }
 
-    fn result(&self, expected_result: Option<usize>) -> Result<Option<Vec<PoseidonResult>>> {
-        let mut results: Vec<PoseidonResult> = vec![];
+    fn result(&self, _param: Option<PoseidonReadResult>) -> Result<Option<Vec<PoseidonResult>>> {
+        let params = _param.unwrap();
 
-        loop {
-            let num_of_pending_results = self.get_num_of_pending_results()?;
-
-            if results.len() >= expected_result.unwrap() {
-                break;
-            }
+        self.dclient.dma_read_into(
+            self.dclient.cfg.dma_baseaddr,
+            DMA_RW::OFFSET,
+            params.result_store_buffer,
+        );
 
-            let res = self.get_raw_results(num_of_pending_results)?;
+        assert_eq!(
+            params.result_store_buffer.len(),
+            params.expected_result * 64
+        );
 
-            let mut result = PoseidonResult::parse_poseidon_hash_results(res);
-            results.append(&mut result);
-        }
+        let results =
+            PoseidonResult::parse_poseidon_hash_results(params.result_store_buffer.get().to_vec());
```

Comment:
Why do we need this new `PoseidonReadResult` struct? Can't we build the buffer in the `result` function based on the `expected_result` size?